### PR TITLE
library: Add Email Entry

### LINKED ISSUE #329: Adds Email Demo
https://libportal.org/method.Portal.compose_email.html

### PR DESCRIPTION
**Pull request by SoNiC-HeRE**
Closes #329

### CODE REVIEW

**Comment by SoNiC-HeRE** [comment]
@andyholmes  Can you please provide an early review for the email demo?
As of last time it seemed to work for everyone else, so just wanna make sure if it's a permission error or something else

**Comment by sonnyp** [comment]
Please make sure to use `Gio._promisify`

**Comment by SoNiC-HeRE** [comment]
> Please make sure to use `Gio._promisify` 

Yes , i was just making sure if it was a problem on my side

File: src/Library/demos/Email/main.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+Gio._promisify(Xdp.Portal.prototype, "compose_email", "compose_email_finish");
+
+async function sendEmail() {
+  try {
+    const result = await portal.compose_email(
+      parent,
+      ["sonicworks05@gmail.com"],
+      ["test@gmail.com"],
+      null,
+      "Demo Message",
```
Comment (sonnyp) [inline src/Library/demos/Email/main.js:21]: can you add a comment next to each line with the name of the argument so it's clear what is what?

For example here

```suggestion
      "Demo Message", // subject
```

File: src/Library/demos/Email/main.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+const picture = workbench.builder.get_object("picture");
+
+Gio._promisify(Xdp.Portal.prototype, "compose_email", "compose_email_finish");
+
+async function sendEmail() {
+  try {
+    const result = await portal.compose_email(
+      parent,
+      ["sonicworks05@gmail.com"],
+      ["test@gmail.com"],
```
Comment (sonnyp) [inline src/Library/demos/Email/main.js:19]: Please don't use your personal data in examples.

```suggestion
      ["example@example.com"], // addresses
      null, // cc
```

File: src/Library/demos/Email/main.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    if (result) {
+      console.log("Email app opened");
+    }
+  } catch (err) {
+    logError(err);
+  }
+}
+
+button.connect("clicked", sendEmail);
```
Comment (sonnyp) [inline src/Library/demos/Email/main.js:36]: I have updated our practice a bit here see https://github.com/sonnyp/Workbench/pull/339

Please remove the try catch from sendEmail to make the function simpler and do the following instead

```suggestion
button.connect("clicked", () => {
  onClicked().catch(logError);
));
```

File: src/Library/demos/Email/main.js (hunk 2 above)
Comment (sonnyp) [inline src/Library/demos/Email/main.js:19]: Please don't use your personal data in examples.

```suggestion
      ["example@example.com"], // addresses
      null, // cc
```

File: src/Library/demos/Email/main.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+async function sendEmail() {
+  try {
+    const result = await portal.compose_email(
+      parent,
+      ["sonicworks05@gmail.com"],
+      ["test@gmail.com"],
+      null,
+      "Demo Message",
+      "Demo Content",
```
Comment (sonnyp) [inline src/Library/demos/Email/main.js:22]: I would suggest "Email from Workbench" as subject and "Hello World!" as body

File: src/Library/demos/Email/main.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "Email",
+  "category": "platform",
+  "description": "Compose an email message",
```
Comment (sonnyp) [inline src/Library/demos/Email/main.json:4]: remember to write description from the perspective of the developer

```suggestion
  "description": "Trigger an email message",
```

**Review by sonnyp — COMMENTED** [review]
It would be nice to add a text input inviting the user to enter their email address and use it for the `addresses` arg.

This way, when the user can test by sending the email to themselves directly. 

wdyt?

File: src/Library/demos/Email/main.blp [hunk 6]
```diff
@@ -0,0 +1,25 @@
+using Gtk 4.0;
+using Adw 1;
+
+Adw.StatusPage {
+  title: "Email";
+  description: _("Compose an email");
```
Comment (sonnyp) [inline src/Library/demos/Email/main.blp:6]:
```suggestion
  description: _("Trigger an email");
```

**Comment by SoNiC-HeRE** [comment]
> It would be nice to add a text input inviting the user to enter their email address and use it for the `addresses` arg.
> 
> This way, when the user can test by sending the email to themselves directly.
> 
> wdyt?

Sure, it's a good idea

**Comment by SoNiC-HeRE** [comment]
Made the following changes
- Used a Gtk.Entry to take email from user
- Commented various properties of portal.compose_email method
- Implemented an email validator
- Refined Code
- Removed async handlers

**Comment by SoNiC-HeRE** [comment]
Made the following changes
- Used a Gtk.Entry to take email from user
- Commented various properties of portal.compose_email method
- Implemented an email validator
- Refined Code
- Removed async handlers

**Review by sonnyp — APPROVED** [review]
Well done.

I made minor changes https://github.com/sonnyp/Workbench/pull/330/commits/a3966fbab420ef092b9d60dada7e211750842147

Make sure to use the name of arguments as defined by the API documentation next time.

**Pull request merged**